core: Ensure `quotes_required` and `quotes_supported` for SEP-31 are handled

erika-sdf opened the pull request:
<!-- If you're making a doc PR or something tiny where the below is irrelevant, delete this
template and use a short description, but in your description aim to include both what the
change is, and why it is being made, with enough context for anyone to understand. -->

<details>
  <summary>PR Checklist</summary>

### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

Add support for SEP-31 `quotes_required` and `quotes_supported`.

### Why

The anchor platform should reject invalid configuration requests  for these values should.

### Known limitations

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Something went wrong with PR preview build please check**

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

erika-sdf commented:
Depends on https://github.com/stellar/stellar-anchor-tests/pull/90/files

marcelosalloum commented on core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:
```diff
@@ -326,9 +326,13 @@ void validatePatchTransactionFields(Sep31Transaction txn, Sep31PatchTransactionR
+    }
+
     // Check if quote is provided.
     if (request.getQuoteId() == null) {
```
👀  Can you also add a condition where `quoteSupported == false` also returns early?

```suggestion
    if (!quoteSupported || request.getQuoteId() == null) {
```

marcelosalloum replied:
Also, could you plz add a test for this new condition?

marcelosalloum commented on core/src/test/resources/test_assets.json.quotes_not_supported:
```diff
@@ -0,0 +1,208 @@
+{
+  "assets": [
+    {
+      "schema": "stellar",
+      "code": "USDC",
```
💡  It seems like you're only using this file in one place, and it's only used to test if the validation fails when quotes are marked as required but not supported. With that in mind, you can probably reduce this file a lot, by leaving just one asset and only the configurations you need for the test. WDYT?

We already have a few similar files in the project and having all configs in all files is a bit hard to maintain, so I'd go leaner with just what you need for this test.

Also, the file name could be more accurate, with something like `test_assets.json.quotes_required_but_not_supported`.

marcelosalloum reviewed (commented):
Congrats, the code and tests are pretty thorough 🎉.

I've added a couple of comments below, one with a missing validation and another with a suggestion 💡.

lijamie98 commented on core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:
```diff
@@ -326,9 +326,13 @@ void validatePatchTransactionFields(Sep31Transaction txn, Sep31PatchTransactionR
   }
 
-  void preValidateQuote() throws AnchorException {
+  void preValidateQuote(boolean isQuotesRequired) throws AnchorException {
```
`assetInfo` is available by calling `Context.get().getAsset();` WDYT if we remove `boolean isAuotesRequired`? 

`
void preValidateQuote() throws AnchorException {
    AssetInfo assetInfo = Context.get().getAsset();
    boolean isQuotesRequired = assetInfo.getSep31().isQuotesRequired();
....
}

`

erika-sdf replied:
Thanks for the suggestion, I've made this change

lijamie98 commented on core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:
```diff
@@ -99,7 +99,7 @@ public Sep31PostTransactionResponse postTransaction(
     }
     validateRequiredFields(assetInfo.getCode(), request.getFields().getTransaction());
```
WDYT if we also remove the arguments passing to `validateRequiredFields` if possible to make the style more consistent.  But feel free not to take care of this at this moment.

erika-sdf replied:
This has been addressed, as well as adding a new call to `AssetService` to get `AssetInfo` with only the asset code.

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

erika-sdf commented:
> Congrats, the code and tests are pretty thorough 🎉.
> 
> I've added a couple of comments below, one with a missing validation and another with a suggestion 💡.

Thanks for the review. I've addressed the changes below.

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

lijamie98 commented on core/src/main/java/org/stellar/anchor/asset/JsonAssetService.java:
```diff
@@ -19,12 +19,22 @@ public List<AssetInfo> listAllAssets() {
+      return getAsset(code);
+    }
+    for (AssetInfo asset : assets.assets) {
+      if (asset.getCode().equals(code) && issuer.equals(asset.getIssuer())) {
```
💡 nit: Can we use the safer version of comparison in case `asset.code()` is null? 
```java
if (code.equals(asset.getCode()) && issuer.equals(asset.getIssuer())) {
```

lijamie98 reviewed (approved):
LGTM.

marcelosalloum commented on core/src/main/java/org/stellar/anchor/asset/JsonAssetService.java:
```diff
@@ -19,12 +19,22 @@ public List<AssetInfo> listAllAssets() {
-  public AssetInfo getAsset(String code, String issuer) {
+  public AssetInfo getAsset(String code) {
     for (AssetInfo asset : assets.assets) {
       if (asset.getCode().equals(code)) {
-        if (issuer == null || issuer.equals(asset.getIssuer())) {
           return asset;
-        }
+      }
+    }
+    return null;
+  }
+
+  public AssetInfo getAsset(String code, String issuer) {
+    if (issuer == null) {
+      return getAsset(code);
+    }
```
nit: the design I see more often in libraries and SDKs is for the method with fewer parameters to call the one with the most parameters, for instance:

```java
public AssetInfo getAsset(String code) {
  return getAsset(code, null);
}
```

Nothing wrong with your design though, we can move on with that.

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Reference Server Preview is available here:**<br>https://anchor-ref-pr319.previews.kube001.services.stellar-ops.com/<br>**SEP Server Preview is available here:**<br>https://anchor-sep-pr319.previews.kube001.services.stellar-ops.com/
